openshift/oadp-operator
# Adds kopia/restic uploader type

Implements #1101 part of the #1046 

Testing instructions:
1. Uninstall previous version of OADP
2. Ensure you are logged as an admin
  ```shell
  $ oc whoami
  system:admin
  ```
4. Remove left over CRD's:
  ```shell
  $ oc delete crd dataprotectionapplications.oadp.openshift.io volumesnapshotrestores.datamover.oadp.openshift.io volumesnapshotbackups.datamover.oadp.openshift.io cloudstorages.oadp.openshift.io
  ```

5a. You may want to use the following patch to not install or symlink podman to docker:

  ```shell
  $ patch -p1 < /below/patch/file/location.diff
  ```

```diff
index 94cfa11..bf250a4 100644
--- a/Makefile
+++ b/Makefile
@@ -212,10 +212,10 @@ run: manifests generate fmt vet ## Run a controller from your host.
 # related enhancements that may remove the need to manually install qemu-user-static https://bugzilla.redhat.com/show_bug.cgi?id=2061584
 DOCKER_BUILD_ARGS ?= --platform=linux/amd64
 docker-build: ## Build docker image with the manager.
-	docker build -t $(IMG) . $(DOCKER_BUILD_ARGS)
+	podman build -t $(IMG) . $(DOCKER_BUILD_ARGS)
 
 docker-push: ## Push docker image with the manager.
-	docker push ${IMG}
+	podman push ${IMG}
 
 ##@ Deployment
 
@@ -361,7 +361,7 @@ nullable-crds-config:
 
 .PHONY: bundle-build
 bundle-build: ## Build the bundle image.
-	docker build -f bundle.Dockerfile -t $(BUNDLE_IMG) . $(DOCKER_BUILD_ARGS)
+	podman build -f bundle.Dockerfile -t $(BUNDLE_IMG) . $(DOCKER_BUILD_ARGS)
 
 .PHONY: bundle-push
 bundle-push: ## Push the bundle image.
```

5b. Build and deploy OADP from this PR:
  ```shell
  $ make build-deploy
  $ make deploy-olm
  ```

6. Create DPA, ensure to choose `restic` or `kopia` and ensure to use `unsupportedOverrides` with provided `veleroImageFqin`

Make sure to specify the `backupLocations` accordingly to your cluster configuration and `s3` bucket

Part of the DPA for restic:
  ```yaml
  apiVersion: oadp.openshift.io/v1alpha1
  kind: DataProtectionApplication
  metadata:
    name: dpa-sample
  spec:
    configuration:
      nodeAgentConfig:
        enable: true
        uploaderType: restic
  [...]
    unsupportedOverrides:
      veleroImageFqin: 'docker.io/sseago/velero:kopia'
  ```
And for Kopia
  ```yaml
  apiVersion: oadp.openshift.io/v1alpha1
  kind: DataProtectionApplication
  metadata:
    name: dpa-sample
  spec:
    configuration:
      nodeAgentConfig:
        enable: true
        uploaderType: kopia
  [...]
    unsupportedOverrides:
      veleroImageFqin: 'docker.io/sseago/velero:kopia'
  ```

8. Create `Backup` object to trigger backup, ensure to include `defaultVolumesToFsBackup: true`
  ```yaml
  apiVersion: velero.io/v1
  kind: Backup
  metadata:
    name: mysql-persistent
    namespace: openshift-adp
  spec:
    includedNamespaces:
    - mysql-persistent
    storageLocation: dpa-sample
    defaultVolumesToFsBackup: true
    ttl: 720h0m0s
  ```
